extensions/exchanges/bitfinex/exchange.js

```diff
@@ -56,201 +248,163 @@ module.exports = function container (get, set, clear) {
+      if (Object.keys(ws_balance).length === 0) { return retry('getBalance', opts, cb) }
+
+      balance = {}
+      balance.currency      = n(ws_balance[opts.currency].balance).format('0.00000000')
+      balance.asset         = n(ws_balance[opts.asset].balance).format('0.00000000')
+
+      balance.currency_hold = ws_balance[opts.currency].available ? n(ws_balance[opts.currency].balance).subtract(ws_balance[opts.currency].available).format('0.00000000') : n(0).format('0.00000000')
+      balance.asset_hold    = ws_balance[opts.asset].available    ? n(ws_balance[opts.asset].balance).subtract(ws_balance[opts.asset].available).format('0.00000000')       : n(0).format('0.00000000')
+
+      cb(null, balance)
     },
     
     getQuote: function (opts, cb) {
-      var func_args = [].slice.call(arguments)
-      var client = publicClient()
-      var pair = 't' + joinProduct(opts.product_id)
-      client.ticker(pair, function (err, body) {
-        if (err) return retry('getQuote', func_args, err)
-        cb(null, { bid : String(body.BID), ask : String(body.ASK) })
-      })
+      cb(null, { bid : String(ws_ticker.BID), ask : String(ws_ticker.ASK) })
     },
     
     cancelOrder: function (opts, cb) {
-      var client = authedClient()
-      client.cancel_order(opts.order_id, function (err, body) {
-        if (err) return(err)
-        cb()
-      })
+      order = ws_orders['~' + opts.order_id]
+      ws_orders['~' + opts.order_id].reject_reason = "zenbot cancel"
+
+      var ws_cancel_order = [
+        0,
+        'oc',
+        null,
+        {
+          id: order.bitfinex_id
+        }
+      ]
+
+      client.send(ws_cancel_order)
+      cb()
     },
     
-    buy: function (opts, cb) {
-      var client = authedClient()
+    trade: function (action, opts, cb) {
+      client = authedClientWs();
+
+      var cid = Math.round(((new Date()).getTime()).toString() * Math.random())
+      var symbol = 't' + joinProduct(opts.product_id)
+      var amount = action === 'buy' ? opts.size : opts.size * -1
+      var price = opts.price
+
       if (opts.order_type === 'maker' && typeof opts.type === 'undefined') {
-        opts.type = 'exchange limit'
+        opts.type = 'EXCHANGE LIMIT'
       }
       else if (opts.order_type === 'taker' && typeof opts.type === 'undefined') {
-        opts.type = 'exchange market'
+        opts.type = 'EXCHANGE MARKET'
       }
       if (typeof opts.post_only === 'undefined') {
         opts.post_only = true
       }
-      var symbol = joinProduct(opts.product_id)
-      var amount = opts.size
-      var price = opts.price
-      var exchange = 'bitfinex'
-      var side = 'buy'
       var type = opts.type
-      var is_hidden = false
       var is_postonly = opts.post_only
-      var params = {
-        symbol,
-        amount,
-        price,
-        exchange,
-        side,
-        type,
-        is_hidden,
-        is_postonly
+
+      var order = {
+        id: cid,
+        bitfinex_id: null,
```

Comment:
since we are using WS API, we immediately get order ID via 'os' or 'hos'.
I also exchanged `bitfinex_id `with `opts.order_id` and it canceled the order.